File: epi_judge_python/real_square_root.py
from test_framework import generic_test


def square_root(x):
    # TODO - you fill in here.
    if x < 1.0:
        low = x
        high = 1.0
    else:
        low = 1.0
        high = x

    while low < high:
        mid = low + (high - low) * 0.5
        print("low: {}, mid: {}, high: {}".format(low, mid, high))
        if mid*mid > x:
            high = mid 
        elif mid*mid < x:
            low = mid 
        elif mid*mid == x:
            return mid

    return low


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("real_square_root.py",
                                       'real_square_root.tsv', square_root))

File: epi_judge_java/epi/RealSquareRoot.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class RealSquareRoot {
    @EpiTest(testDataFile = "real_square_root.tsv")

    public static double squareRoot(double x) {
        // TODO - you fill in here.

        final double EP = 0.000001;
        double left, right;
        if (x < 1.0) {
            left = x;
            right = 1.0;
        }
        else {
            left = 1.0;
            right = x;
        }
        
        while (((left - right) / right) < -EP) {
            double mid = left + (right - left) * 0.5;
            double midSquared = mid * mid;
            double diff = (midSquared - x) / x;

            if (diff > EP) {
                right = mid;
            }
            else if (diff < -EP) {
                left = mid;
            }
            else {
                return mid;
            }
        }

        // while (compare(left, right) == Ordering.SMALLER) {
        //     double mid = left + (right - left) * 0.5;
        //     double midSquared = mid * mid;

        //     if (compare(midSquared, x) == Ordering.LARGER) {
        //         right = mid;
        //     }
        //     else if (compare(midSquared, x) == Ordering.SMALLER) {
        //         left = mid;
        //     }
        //     else {
        //         return mid;
        //     }
        // }
        return left; // Time = O(log (x/s)) (s = tolerance)
    }

    private static enum Ordering { SMALLER, EQUAL, LARGER }

    private static Ordering compare(double a, double b) {
        final double EPSILON = 0.000001;

        double diff = (a - b) / b;
        return diff < -EPSILON ?
            Ordering.SMALLER :
            (diff > EPSILON ? Ordering.LARGER : Ordering.EQUAL);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                .runFromAnnotations(args, "RealSquareRoot.java",
                    new Object() {}.getClass().getEnclosingClass())
                .ordinal());
    }
}
